src/coprocessor/dag/expr/builtin_math.rs

```diff
@@ -330,4 +337,34 @@ mod test {
             assert_eq!(got, exp);
         }
     }
+
+    #[test]
+    fn test_sin() {
+        let tests = vec![
+            (ScalarFuncSig::Sin, Datum::F64(0.0_f64), Datum::F64(0.0_f64)),
+            (
+                ScalarFuncSig::Sin,
+                Datum::F64(f64::consts::PI / 4.0_f64),
+                Datum::F64(0.7071067811865475_f64),
+            ),
+            (
+                ScalarFuncSig::Sin,
+                Datum::F64(f64::consts::PI / 2.0_f64),
+                Datum::F64(1.0_f64),
+            ),
+            (
+                ScalarFuncSig::Sin,
+                Datum::F64(f64::consts::PI),
+                Datum::F64(0.00000000000000012246467991473532_f64),
```

Comment:
**sin(π)** should be equal to 0. While it is clear why the actual result is not precisely equal to 0, I'm afraid that hard-coding it as this particular value may break the build on other architectures.

Reply:
@zz-jason any suggestion?

Reply:
How about replace the result to `sin(pi())`.

Another thing, in MySQL, `sin(pi())` returns a float-point value nearly to zero, not exactly zero:
```sql

MySQL(localhost:3306) > select sin(pi());
+------------------------+
| sin(pi())              |
+------------------------+
| 1.2246467991473532e-16 |
+------------------------+
1 row in set (0.00 sec)
```

And TiDB behaves the same with MySQL and returns almost the same small value:
```sql
TiDB(localhost:4000) > select sin(pi());
+------------------------+
| sin(pi())              |
+------------------------+
| 1.2246467991473515e-16 |
+------------------------+
1 row in set (0.00 sec)
```

Reply:
@vkorenev I assume sin(pi) = 0 is in theory, in computer float calculation it cannot be precise. And thats' also why you normally don't do float comparison with ==.

Should I switch to `sin(pi())`? however I feel if we are going to switch to this, then if rust's implementation of `f64::sin()` changes, then this test will still pass, but we wouldn't know by then the tikv will give a different results than Mysql. (Though this is very very unlikely to happen)

Please let me know what to do now.

Reply:
I think the reason why all the programs above return the same result is that **sin** is computed by the same CPU instruction on the same processor family. I think the result may be slightly different on some other processor/arch (AMD?, ARM?) or even on another Intel mircoarchitecture.
Anyway, I'd rather compare the computed value with the mathematically correct result +- some epsilon.